app/src/main/java/io/homeassistant/companion/android/widgets/graph/GraphWidget.kt

```diff
@@ -0,0 +1,388 @@
+            }
+
+            axisLeft.apply {
+                setDrawGridLines(true)
+                textColor = dynTextColor
+                textSize = 12F
+            }
+
+            axisRight.apply {
+                setDrawGridLines(false)
+                setDrawLabels(false)
+            }
+
+            legend.apply {
+                isEnabled = true
+                textColor = dynTextColor
+                textSize = 12F
+                verticalAlignment = Legend.LegendVerticalAlignment.BOTTOM
+                horizontalAlignment = Legend.LegendHorizontalAlignment.RIGHT
+                orientation = Legend.LegendOrientation.HORIZONTAL
+                setDrawInside(false)
+            }
+
+            description = Description().apply {
+                text = "$timeRange h"
+                textColor = dynTextColor
+            }
+
+            legend.isEnabled = true
+            description.isEnabled = true
+        }
+
+        val mainGraphColor = ContextCompat.getColor(context, commonR.color.colorPrimary)
+
+        val dataSet = LineDataSet(entries, label).apply {
+            color = mainGraphColor
+            lineWidth = 2F
+            circleRadius = 1F
+            setDrawCircleHole(false)
+            setCircleColor(mainGraphColor)
+            mode = LineDataSet.Mode.CUBIC_BEZIER
+            setDrawCircles(true)
+            setDrawValues(false)
+        }
+
+        lineChart.data = LineData(dataSet)
+
+        lineChart.layout(0, 0, width, height)
+
+        return lineChart
+    }
+
+    override suspend fun getAllWidgetIdsWithEntities(context: Context): Map<Int, Pair<Int, List<String>>> =
+        repository.getAllFlow()
+            .first()
+            .associate { it.id to (it.serverId to listOf(it.entityId)) }
+
+    private suspend fun resolveTextToShow(
+        serverId: Int,
+        entityId: String?,
+        suggestedEntity: Entity<Map<String, Any>>?
+    ): ResolvedText {
+        var entity: Entity<Map<String, Any>>? = null
+        var entityCaughtException = false
+        try {
+            entity = if (suggestedEntity != null && suggestedEntity.entityId == entityId) {
+                suggestedEntity
+            } else {
+                entityId?.let { serverManager.integrationRepository(serverId).getEntity(it) }
+            }
+        } catch (e: Exception) {
+            Log.e(TAG, "Unable to fetch entity", e)
+            entityCaughtException = true
+        }
+
+        return ResolvedText(entity?.friendlyName, entityCaughtException)
+    }
+
+    private suspend fun fetchHistory(appWidgetId: Int, serverId: Int, entityId: String, fromMillis: Long, toMillis: Long) {
+        val entitiesStatesList: List<List<Entity<Map<String, Any>>>>?
+        try {
+            entitiesStatesList = serverManager.integrationRepository(serverId)
+                .getHistory(
+                    listOf(entityId),
+                    fromMillis,
+                    toMillis
+                )
```

Comment:
Any particular reason for not using `minimalResponse` in the request here? That should significantly reduce the data needed, and you're not using any of the other data.